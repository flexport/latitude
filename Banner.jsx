/**
 * TEAM: frontend_infra
 * @flow strict
 */

/* eslint-disable flexport/no-unused-aphrodite-styles */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Icon, {type IconNames} from "./Icon";
import invariant from "./tools/invariant";
import IconButton from "./button/IconButton";
import latitudeColors, {type Color} from "./colors";

import Text from "./Text";

type Intent = "default" | "default-light" | "warning" | "danger";

type Props = {|
  /**
   * defines the visual style which conveys the level of importance / urgency
   * to the user
   */
  +intent?: Intent,
  /** the name of the icon to be displayed on the left of the banner */
  +iconName?: IconNames,
  /** the message that will appear at the center of the banner */
  +message: string | React.Element<typeof Text>,
  /** additional content to display in the banner below the message */
  +additionalContent?: React.Node,
  /**
   * The call to action button that will appear below the banner message.
   * This button should have intent "basic" and kind "blank"
   */
  +ctaButton?: React.Node,
  /**
   * Called when the close button is pressed. If an onClose
   * isn't provided, the close button will not appear
   */
  +onClose?: () => void,
|};

/**
 * @category Data Display
 * @short Spans the entire width of its parent and displays a informing message
 * @brandStatus V2
 * @status Beta
 *
 * The banner component is used to provide feedback to the user such as when
 * an app update has occured or when there are errors returned from the server.
 * Banners can also be used to display passive messages such as when times are
 * being displayed in a separate timezone or when the user is impersonating
 * a client.
 */
export default function Banner({
  intent = "default",
  iconName,
  message,
  additionalContent,
  ctaButton,
  onClose,
}: Props) {
  invariant(
    !(intent === "danger" && ctaButton),
    "Danger Banners cannot have a call to action button"
  );

  const messageContent = message === "string" ? <Text>message</Text> : message;

  const iconColor = getIconColor(intent);

  const ctaButtonContent =
    ctaButton != null ? (
      <div className={css(styles.ctaButtonWrapper)}>{ctaButton}</div>
    ) : null;

  const additionalContentNode =
    additionalContent != null ? (
      <div className={css(styles.additionalContentWrapper)}>
        {additionalContent}
      </div>
    ) : null;

  let closeNode = null;
  if (onClose) {
    closeNode = (
      <div className={css(intent === "danger" && styles.closeWrapperDanger)}>
        <IconButton
          iconName="cancel"
          onClick={onClose}
          kind="blank"
          intent="none"
          type="button"
          size="l"
          height={{type: "customDontUse", height: 20}}
        />
      </div>
    );
  }

  return (
    <div
      className={css(
        styles.container,
        styles[intent],
        additionalContent != null && styles.closeButtonWrapper
      )}
    >
      <div className={css(styles.contentWrapper)}>
        <div className={css(styles.iconWrapper)}>
          {iconName && <Icon size="s" iconName={iconName} color={iconColor} />}
        </div>
        <div className={css(styles.content)}>
          {messageContent}
          {additionalContentNode}
          {ctaButtonContent}
        </div>
      </div>
      {closeNode}
    </div>
  );
}

const getIconColor = (intent: Intent): Color => {
  if (intent === "danger") {
    return "white";
  } else if (intent === "warning") {
    return "red40";
  }
  return "grey60";
};

const styles = StyleSheet.create({
  container: {
    display: "flex",
    flexDirection: "row",
    borderRadius: "0",
    alignItems: "flex-start",
  },
  closeButtonWrapper: {
    alignItems: "baseline",
  },
  contentWrapper: {
    display: "flex",
    flexDirection: "row",
    flex: 1,
  },
  iconWrapper: {
    marginRight: "20px",
  },
  closeWrapperDanger: {
    ":nth-child(1n) > button > span > svg": {
      fill: latitudeColors.white,
      animationDuration: "0",
    },
    ":nth-child(1n) > button > span > svg :hover": {
      fill: latitudeColors.white,
    },
  },
  content: {
    display: "flex",
    flexDirection: "column",
    flex: "1",
    marginRight: "20px",
    lineHeight: "20px",
  },
  additionalContentWrapper: {
    marginTop: "8px",
  },
  ctaButtonWrapper: {
    paddingTop: "8px",
  },

  default: {
    background: latitudeColors.grey10,
    padding: "18px",
    border: `2px solid ${latitudeColors.grey20}`,
  },
  "default-light": {
    background: latitudeColors.white,
    padding: "18px",
    border: `2px solid ${latitudeColors.grey20}`,
  },
  warning: {
    border: `2px solid ${latitudeColors.red40}`,
    padding: "18px",
    background: latitudeColors.white,
  },
  danger: {
    background: latitudeColors.red40,
    padding: "20px",
    color: latitudeColors.white,
  },
});
