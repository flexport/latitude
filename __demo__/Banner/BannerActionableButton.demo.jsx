/**
 * TEAM: frontend_infra
 * @flow
 */

import React from "react";

import Group from "../../Group";
import Banner from "../../Banner";
import Button from "../../button/Button";

/**
 * @title Banner With Actionable Button
 * @description Banners can become actionable with the addition of a button.
 */
export default function BannerActionableButton() {
  return (
    <Group flexDirection="column" gap={40}>
      <Banner
        iconName="attention"
        message="You cannot edit an automatic On Duty Log. Youâ€™ll need to create a new log."
        onClose={() => {}}
        ctaButton={
          <Button label="Create a new log" kind="blank" intent="basic" />
        }
      />
    </Group>
  );
}
