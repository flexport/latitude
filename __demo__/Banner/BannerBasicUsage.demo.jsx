/**
 * TEAM: frontend_infra
 * @flow
 */
import React from "react";
import Group from "../../Group";
import Banner from "../../Banner";

/**
 * @title Basic usage
 * @description Banners can use different intents for different use cases.
 */
export default function BannerBasicUsage() {
  return (
    <Group flexDirection="column" gap={40}>
      <Group flexDirection="column">
        {'Intent "Default":'}
        <Banner
          iconName="attention"
          message="You cannot edit an automatic On Duty Log. You’ll need to create a new log."
          onClose={() => {}}
        />
      </Group>
      <Group flexDirection="column">
        {'Intent "Warning":'}
        <Banner
          iconName="attention"
          message="You cannot edit an automatic On Duty Log. You’ll need to create a new log."
          intent="warning"
          onClose={() => {}}
        />
      </Group>
      <Group flexDirection="column">
        {'Intent "Danger":'}
        <Banner
          iconName="attention"
          message="You cannot edit an automatic On Duty Log. You’ll need to create a new log."
          intent="danger"
          onClose={() => {}}
        />
      </Group>
    </Group>
  );
}
