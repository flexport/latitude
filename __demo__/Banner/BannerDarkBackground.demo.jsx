/**
 * TEAM: frontend_infra
 * @flow
 */

import React, {type Node} from "react";
import {css, StyleSheet} from "aphrodite";

import {padding} from "../../styles";
import colors from "../../colors";
import Banner from "../../Banner";

/**
 * @title Banner without close button
 * @description Ommitting the closeButton callback will cause the close button to be omitted.
 */
export default function BannerWithoutCloseButton() {
  type Props = {|+children: Node|};

  const styles = StyleSheet.create({bg: {backgroundColor: colors.grey10}});

  const ComponentWithDarkBackground = ({children}: Props) => (
    <div className={css(padding.a.xl, styles.bg)}>{children}</div>
  );

  return (
    <ComponentWithDarkBackground>
      <Banner
        iconName="flag"
        intent="default-light"
        message="You can distinguish this banner from its background!"
      />
    </ComponentWithDarkBackground>
  );
}
