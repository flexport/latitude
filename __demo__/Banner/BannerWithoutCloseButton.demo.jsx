/**
 * TEAM: frontend_infra
 * @flow
 */

import React from "react";

import Group from "../../Group";
import Banner from "../../Banner";
import Button from "../../button/Button";

/**
 * @title Banner without close button
 * @description Ommitting the closeButton callback will cause the close button to be omitted.
 */
export default function BannerWithoutCloseButton() {
  return (
    <Group flexDirection="column" gap={40}>
      <Banner
        iconName="attention"
        intent="warning"
        message="You cannot edit an automatic On Duty Log. Youâ€™ll need to create a new log."
        ctaButton={
          <Button label="Create a new log" kind="blank" intent="basic" />
        }
      />
    </Group>
  );
}
