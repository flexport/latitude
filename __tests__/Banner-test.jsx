/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {mount} from "enzyme";
import Banner from "../Banner";

describe("Banner", () => {
  it("matches snapshot", () => {
    const banner = mount(<Banner message="I am a banner" />);
    expect(banner).toMatchSnapshot();
  });
});
