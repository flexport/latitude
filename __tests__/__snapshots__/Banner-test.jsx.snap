// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner matches snapshot 1`] = `
<Banner
  message="I am a banner"
>
  <div
    className="container_1m2wvan-o_O-default_1tkyt1i"
  >
    <div
      className="contentWrapper_pmg5nu"
    >
      <div
        className="iconWrapper_kh8dwt"
      />
      <div
        className="content_1kq4jtl"
      >
        I am a banner
      </div>
    </div>
  </div>
</Banner>
`;
